modules/lib_vnr/src/features/vnr_features_priv.c

```diff
@@ -0,0 +1,169 @@
+{
+#if (VNR_FD_FRAME_LENGTH < (VNR_MEL_FILTERS*VNR_PATCH_WIDTH))
+    #error ERROR squared_mag_data memory not enough for reuse as normalised_patch
+#endif
+    bfp_s32_init(normalised_patch, normalised_patch_data, 0, VNR_MEL_FILTERS*VNR_PATCH_WIDTH, 1);
+    //norm_patch = feature_patch - np.max(feature_patch)   
+    bfp_s32_t feature_patch_bfp;
+    bfp_s32_init(&feature_patch_bfp, (int32_t*)&feature_state->feature_buffers[0][0], VNR_LOG2_OUTPUT_EXP, VNR_MEL_FILTERS*VNR_PATCH_WIDTH, 1);
+    float_s32_t max = bfp_s32_max(&feature_patch_bfp);
+    float_s32_t zero = {0, 0};
+    float_s32_t neg_max = float_s32_sub(zero, max);
+    bfp_s32_add_scalar(normalised_patch, &feature_patch_bfp, neg_max);
+}
+void vnr_priv_mel_compute(float_s32_t *filter_output, const bfp_complex_s32_t *X) {
+#if VNR_MEL_FILTERS != AUDIO_FEATURES_NUM_MELS
+    #error VNR_MEL_FILTERS not the same as AUDIO_FEATURES_NUM_MELS
+#endif
+    // Calculate squared magnitude spectrum
+    int32_t DWORD_ALIGNED squared_mag_data[VNR_FD_FRAME_LENGTH];
+    bfp_s32_t squared_mag;
+    bfp_s32_init(&squared_mag, squared_mag_data, 0, X->length, 0);
+    bfp_complex_s32_squared_mag(&squared_mag, X);
+#if HEADROOM_CHECK
+    headroom_t reported_hr = squared_mag.hr;
+    headroom_t actual_hr = bfp_s32_headroom(&squared_mag);
+    if(reported_hr != actual_hr) {
+        printf("ERROR: bfp_complex_s32_squared_mag(), reported hr %d, actual %d\n",reported_hr, actual_hr);
+        assert(0);
+    }
+#endif
+
+    // Mel filtering
+    unsigned num_bands = AUDIO_FEATURES_NUM_MELS + 1; //Extra band for the 2nd half of the last filter  
+    unsigned num_even_filters = num_bands / 2;
+    unsigned num_odd_filters = AUDIO_FEATURES_NUM_MELS - num_even_filters;
+    int filter_exp = -(31-AUDIO_FEATURES_MEL_HEADROOM_BITS);
+    int filter_hr = AUDIO_FEATURES_MEL_HEADROOM_BITS;
+    for(unsigned i=0; i<num_even_filters; i++) {
+        unsigned filter_start = mel_filter_512_24_compact_start_bins[2*i];
+        unsigned filter_length = mel_filter_512_24_compact_start_bins[2*(i+1)] - filter_start;
+        //Create the bfp for spectrum subset
+        bfp_s32_t spect_subset;
+        bfp_s32_init(&spect_subset, &squared_mag.data[filter_start], squared_mag.exp, filter_length, 0);
+        spect_subset.hr = squared_mag.hr;
+
+        //Create BFP for the filter
+        bfp_s32_t filter_subset;
+        bfp_s32_init(&filter_subset, &mel_filter_512_24_compact_q31[filter_start], filter_exp, filter_length, 0);
+        filter_subset.hr = filter_hr;
+        
+        filter_output[2*i] = float_s64_to_float_s32(bfp_s32_dot(&spect_subset, &filter_subset));
+    }
+
+    bfp_s32_t filter;
+    bfp_s32_init(&filter, &mel_filter_512_24_compact_q31[0], filter_exp, AUDIO_FEATURES_NUM_BINS, 0);
+
+    int32_t odd_band_filters_data[AUDIO_FEATURES_NUM_BINS];
+    bfp_s32_t odd_band_filters;
+    bfp_s32_init(&odd_band_filters, odd_band_filters_data, 0, AUDIO_FEATURES_NUM_BINS, HR_S32(AUDIO_FEATURES_MEL_MAX));
```

Comment:
Don't understand the point of using HR_S32 when it returns headroom. This function expects 0 or 1. If one it will calculate the headroom of this data array, if zero, you'll have to assign it yourself